### ISSUE TITLE
Optional fields not so optional as they seems...

### ISSUE THREAD

**Issue by nolanlawson** [pouchdb/pouchdb#6268]
_From @FlamesoFF on June 8, 2016 21:22_

**Example:**
This code works well:
`db.find({
  selector: {name: 'Mario'},
  fields: ['_id', 'name'],
  sort: ['name']
});`

And this does not:
`db.find({
  selector: {name: 'Mario'}
});`

Error:

> TypeError: Cannot convert undefined or null to object
>     at checkFieldsLogicallySound (https://nolanlawson.github.io/pouchdb-find/dist/pouchdb.find.js:1869:35)
>     at checkIndexMatches (https://nolanlawson.github.io/pouchdb-find/dist/pouchdb.find.js:1893:24)
>     at findMatchingIndexes (https://nolanlawson.github.io/pouchdb-find/dist/pouchdb.find.js:1910:24)
>     at findBestMatchingIndex (https://nolanlawson.github.io/pouchdb-find/dist/pouchdb.find.js:1922:25)
>     at planQuery (https://nolanlawson.github.io/pouchdb-find/dist/pouchdb.find.js:2163:15)
>     at https://nolanlawson.github.io/pouchdb-find/dist/pouchdb.find.js:1663:21

Documentation says:

> fields (Optional) Defines a list of fields that you want to receive. If omitted, you get the full documents.
> 
> sort (Optional) Defines a list of fields defining how you want to sort. Note that sorted fields also have to be selected in the selector.

_Copied from original issue: nolanlawson/pouchdb-find#198_

**Comment by nolanlawson** [comment]
_From @garrensmith on June 9, 2016 9:43_

What error do you get?

**Comment by nolanlawson** [comment]
_From @FlamesoFF on June 9, 2016 17:1_

@garrensmith Refresh page.

**Comment by nolanlawson** [comment]
_From @raydaly on September 20, 2016 19:54_

If the query specifies a sort like this:

``` javascript
{"selector":{"_id":{"$exists":true}},"fields":["_id","city"],"sort":[]}
```

where sort is an empty array, then an error of "Cannot convert undefined or null to object". 

The error is generated when creating "var isDescending" where requestDef.sort is tested. Adding a test of 

``` javascript
requestDef.sort.length>0 &&
```

after "requestDef.sort &&" solves the problem when sort is an empty string. This was discovered when working with Polymer's app-pouchdb-index.

When the fields array is empty in the selector

``` javascript
{"selector":{"_id":{"$exists":true}},"fields":[],"sort":[]}
```

the result is an empty object. This is true even after the above fix with or without sort being an empty array. The problem is when requestDef.fields is tested to see if it exists and not also to see if it is empty. So it was solved by adding the second test to this line 

``` javascript
  if (requestDef.fields && requestDef.fields.length>0) {
```

Unfortuately I just installed Couchdb 2.0 on my local machine and I'm unable to set up the environment for testing as specified in the Contributing doc. So I've not submitted a pull request.
